### ISSUE TITLE
packaging decision

### ISSUE THREAD

**Issue by loomis** [nuvla/nuvla#8]
Make an explicit decision on the packaging format to be used for Nuvla.  In the past, this has been based on RPM packages.  With the current emphasis on containers, the packaging should probably instead focus on containers as well.  An explicit decision about the service packaging should be made and documented.

**Comment by cjdcordeiro** [comment]
I suggest having a look at Docker Apps. It's quite new (haven't tried yet), but seems suitable for this purpose: https://github.com/docker/app

In summary, it lets you bundle your compose file into a single deployable and parameterized file, which can be uploaded to Docker Hub. So it's pretty similar to a RPM.

**Comment by loomis** [comment]
Docker Apps looks like a more flexible way to package multi-component applications, but doesn't appear to offer any real benefit for single services.  It relies on individual images, that we would still need to generate ourselves.  I think that there is a tighter correspondence between the RPM and the Dockerfile used to create the image.

**Comment by cjdcordeiro** [comment]
well depends on whether you deploy from a Docker image or Dockerfile. If the latter, we don't really build images anymore, and maintain raw Dockerfiles which can be tagged directly in GH.

**Comment by loomis** [comment]
@cjdcordeiro I don't see how that really helps.  Unless I'm missing something (always a possibility) none of the usual ways of running containers takes a Dockerfile+context as an input. That is, either the image needs to be prebuilt by someone or the user has to build the image himself before running it. 

Aside from that, most of our services involve a significant amount of custom software and dependencies, that is, a Dockerfile isn't sufficient. If the Dockerfile+context was our standard packaging technique, then the GitHub repository would have to have the Dockerfile itself plus all our code and dependencies.  Aside from being cumbersome, git is not very adapted for managing large numbers of binary files, especially if they change frequently. 

We could, of course, avoid the context by packaging our software in standard OS packages (e.g. RPM). But then we have to build those packages, maintain a repository for them, ensure that they work as traditional services, etc.  That means that we'd need to continue doing everything we currently do in the build plus also maintain Dockerfile descriptions of our services.

**Comment by cjdcordeiro** [comment]
> That is, either the image needs to be prebuilt by someone or the user has to build the image himself before running it.

no, the image can be built automatically by Docker at deployment time. check the `build` attribute in compose.

I'm not sure I get the issue but maybe we can discuss it further later on. If I got it right though, if there is context/binaries that needs to be built and recursively passed to the final container, then you are looking at a multistaged Dockerfile

**Comment by loomis** [comment]
Thanks for pointing out that option.  Indeed it would work fine for local build, although there is still the problem with storing lots of artifacts next to the compose file in git.

There is this note with the "build" option in the compose v3 reference:

> Note: This [build] option is ignored when deploying a stack in swarm mode with a (version 3) Compose file. The docker stack command accepts only pre-built images.

It seems that creating images is the only viable option for remote deployment.

**Comment by konstan** [comment]
IMO, docker run/service/compose/app is orchestration and CM. The initial question was only about packaging and eventually convenience of delivery via deployment (which also includes service startup). Docker image has it all. So, I vote for
* removing RPM packaging
* package service artefacts (jar, war etc.) as Docker images
* (need to see where to use multi-stage Dockerfile or Nexus when dealing with our own core dependencies)
* images should be pre-built by CI (Jenkins) and pushed to registry
* multi-tagging can be used to pass images through test / staging / prod.

**Comment by loomis** [comment]
The decision from this ticket has been summarized in [0002: Packaging](https://github.com/nuvla/nuvla/wiki/0002:-Packaging) on the nuvla/nuvla wiki. Please review and provide comments.

**Comment by cjdcordeiro** [comment]
since this ticket is still open I'll comment here about @konstan 's terminology above. I'd agree in general but I think it is really important to define terms and concepts in advance otherwise we'll end up mixing things in the future.

Let's do a Russian doll :) 

########### Docker 

the software which is at the very core of any container command


######### `docker run`

orchestration of single container applications - **I don't see us using this, at least for now**


####### `docker service`

orchestrations of containers in production. A higher level abstraction of a container application, which allows scaling and load balancing for distributed apps


##### `docker-compose`

orchestration of distributed application bundle (multi-container applications)

### `docker-app`

**packaging and distribution** of cloud native applications. Can also install (_orchestrate_) the apps in an existing COE

 - `docker-app init` packages the app
 - `docker-app render` renders the compose file that will be deployed by docker-compose
 - `docker-app install/uninstall/status` allows to manage the application bundle into an existing infrastructure. It assumes a Docker context exists


---
so in summary, I do believe `docker-app` is a very clean way to package the full software stack, and have it parameterized in such a way that we only need to render the new Docker Image versions on every Nuvla upgrade

**Comment by loomis** [comment]
Final for now.  Further decisions on platform packaging will take place once we have a clearer picture of the landscape and our deployment needs.
